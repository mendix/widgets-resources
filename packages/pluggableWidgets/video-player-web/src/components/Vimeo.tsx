import { Component, createElement, createRef } from "react";
import ReactResizeDetector from "react-resize-detector";
import { fixHeightWithRatio, getRatio, validateUrl } from "../utils/Utils";

export interface VimeoProps {
    url: string;
    autoPlay: boolean;
    loop: boolean;
    muted: boolean;
    aspectRatio?: boolean;
}

export interface VimeoState {
    ratio: number;
}

export class Vimeo extends Component<VimeoProps, VimeoState> {
    private iframe = createRef<HTMLIFrameElement>();
    private readonly handleOnResize = this.onResize.bind(this);
    private handleAttributes = this.getUrlAttributes.bind(this);
    readonly state: VimeoState = {
        ratio: 0
    };

    render(): JSX.Element {
        return (
            <iframe
                className="widget-video-player-iframe"
                src={this.generateUrl(this.props.url)}
                frameBorder="0"
                allow="autoplay; fullscreen"
                allowFullScreen
                ref={this.iframe}
            >
                <ReactResizeDetector
                    handleWidth
                    handleHeight
                    onResize={this.handleOnResize}
                    refreshMode="debounce"
                    refreshRate={100}
                />
            </iframe>
        );
    }

    private onResize(): void {
        if (this.iframe && this.iframe.current && this.props.aspectRatio) {
            if (this.state.ratio) {
                fixHeightWithRatio(this.iframe.current, this.state.ratio);
            } else {
                getRatio(this.props.url).then(ratio => {
                    this.setState({ ratio });
                    if (this.iframe && this.iframe.current) {
                        fixHeightWithRatio(this.iframe.current, this.state.ratio);
                    }
                });
            }
        }
    }

    private generateUrl(url: string): string {
        const attributes = this.handleAttributes();
        try {
            if (url.includes("player.vimeo.com")) {
                return `${url}${attributes}`;
            }

            const urlVimeoSplit = url.split("/");
            if (urlVimeoSplit.length > 0) {
                const id = urlVimeoSplit[urlVimeoSplit.length - 1];
                if (Number(id) > 0 && isFinite(Number(id))) {
                    return `https://player.vimeo.com/video/${id}${attributes}`;
                }
            }
        } catch (e) {
            return url;
        }
        return url;
    }

    private getUrlAttributes(): string {
        let attributes = "?dnt=1";
        attributes += "&autoplay=" + (this.props.autoPlay ? "1" : "0");
        attributes += "&muted=" + (this.props.muted ? "1" : "0");
        attributes += "&loop=" + (this.props.loop ? "1" : "0");
        return attributes;
    }

    // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
    public static canPlay(url: string): boolean {
        return !!url && !!validateUrl(url) && url.indexOf("vimeo.com") > -1;
    }
}
