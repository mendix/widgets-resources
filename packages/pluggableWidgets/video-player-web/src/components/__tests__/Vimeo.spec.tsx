import { createElement } from "react";
import { create } from "react-test-renderer";

import { Vimeo, VimeoProps } from "../Vimeo";
import ReactResizeDetector from "react-resize-detector";

describe("VimeoPlayer Player", () => {
    const defaultProps = {
        url: "http://vimeo.com/123456",
        autoPlay: false,
        muted: false,
        loop: false,
        aspectRatio: false
    };

    const defaulPlayer = (props: VimeoProps): JSX.Element => <Vimeo {...props} />;

    it("should renders correctly", () => {
        const player = create(defaulPlayer(defaultProps)).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with autoplay", () => {
        const player = create(defaulPlayer({ ...defaultProps, autoPlay: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with muted", () => {
        const player = create(defaulPlayer({ ...defaultProps, muted: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with aspectRatio", () => {
        const player = create(defaulPlayer({ ...defaultProps, aspectRatio: true }));
        window.dispatchEvent(new Event("resize"));
        const instance = player.root;
        const sizeDetector = instance.findByType(ReactResizeDetector);

        expect(player.toJSON).toMatchSnapshot();
        expect(sizeDetector).not.toBeNull();
        expect(sizeDetector.props).toHaveProperty("onResize");
    });
});
