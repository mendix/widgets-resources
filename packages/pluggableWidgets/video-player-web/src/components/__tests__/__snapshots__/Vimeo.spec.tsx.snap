// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VimeoPlayer Player should renders correctly 1`] = `
<iframe
  allow="autoplay; fullscreen"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://player.vimeo.com/video/123456?dnt=1&autoplay=0&muted=0&loop=0"
>
  <div />
</iframe>
`;

exports[`VimeoPlayer Player should renders correctly with aspectRatio 1`] = `[Function]`;

exports[`VimeoPlayer Player should renders correctly with autoplay 1`] = `
<iframe
  allow="autoplay; fullscreen"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://player.vimeo.com/video/123456?dnt=1&autoplay=1&muted=0&loop=0"
>
  <div />
</iframe>
`;

exports[`VimeoPlayer Player should renders correctly with muted 1`] = `
<iframe
  allow="autoplay; fullscreen"
  allowFullScreen={true}
  className="widget-video-player-iframe"
  frameBorder="0"
  src="https://player.vimeo.com/video/123456?dnt=1&autoplay=0&muted=1&loop=0"
>
  <div />
</iframe>
`;
